packages/core/types/index.js

```diff
@@ -294,129 +293,78 @@ export type GraphTraversalCallback<TNode, TContext> = (
   traversal: TraversalActions
 ) => ?TContext;
 
-export type NodeId = string;
-
-export type AssetNode = {|id: string, type: 'asset', value: Asset|};
-export type AssetReferenceNode = {|
-  id: string,
-  type: 'asset_reference',
-  value: Asset
-|};
+// Not a directly exported interface.
+interface AssetGraphLike {
+  getDependencyResolution(dependency: Dependency): ?Asset;
+  traverseAssets<TContext>(
+    visit: GraphTraversalCallback<Asset, TContext>
+  ): ?TContext;
+}
 
-export type BundleNode = {|
-  id: string,
-  type: 'bundle',
-  value: Bundle
-|};
+export type MainAssetGraphTraversable =
+  | {|+type: 'asset', value: Asset|}
+  | {|+type: 'dependency', value: Dependency|};
 
-export type BundleGroupNode = {|
-  id: string,
-  type: 'bundle_group',
-  value: BundleGroup
-|};
+// Always read-only.
+export interface MainAssetGraph extends AssetGraphLike {
+  createBundle(asset: Asset): MutableBundle;
+  traverse<TContext>(
+    visit: GraphTraversalCallback<MainAssetGraphTraversable, TContext>
+  ): ?TContext;
+}
 
-export type DependencyNode = {|
-  id: string,
-  type: 'dependency',
-  value: Dependency
-|};
+export interface Bundle extends AssetGraphLike {
+  +id: string;
+  +type: string;
+  +env: Environment;
+  +isEntry: ?boolean;
+  +target: ?Target;
+  +filePath: ?FilePath;
+  +stats: Stats;
+  getBundleGroups(): Array<BundleGroup>;
+  getBundlesInGroup(BundleGroup): Array<Bundle>;
+  getDependencies(asset: Asset): Array<Dependency>;
+  getEntryAssets(): Array<Asset>;
+  getTotalSize(asset?: Asset): number;
+}
 
-export type FileNode = {|id: string, type: 'file', value: File|};
-export type RootNode = {|id: string, type: 'root', value: string | null|};
+export interface MutableBundle extends Bundle {
+  filePath: ?FilePath;
+  isEntry: ?boolean;
+  stats: Stats;
+  merge(Bundle): void;
+  removeAsset(Asset): void;
+}
 
-export type TransformerRequestNode = {|
-  id: string,
-  type: 'transformer_request',
-  value: TransformerRequest
-|};
+export interface RuntimeBundle extends Bundle {
+  addRuntimeAsset({
+    filePath: FilePath,
+    code: string,
+    bundleGroup?: BundleGroup
+  }): Promise<void>;
+}
 
 export type BundleGroup = {
   dependency: Dependency,
   target: ?Target,
   entryAssetId: string
 };
 
-export type Bundle = {|
-  id: string,
-  type: string,
-  assetGraph: AssetGraph,
-  env: Environment,
-  isEntry?: boolean,
-  target?: Target,
-  filePath?: FilePath,
-  stats: Stats
-|};
-
-export type AssetGraphNode =
-  | AssetNode
-  | AssetReferenceNode
-  | DependencyNode
-  | FileNode
-  | RootNode
-  | TransformerRequestNode
-  // Bundle graphs are merged into asset graphs during the bundling phase
-  | BundleGraphNode;
-
-export type BundleGraphNode = BundleNode | BundleGroupNode | RootNode;
-
-export type Edge = {|
-  from: NodeId,
-  to: NodeId
-|};
-
-export type GraphUpdates<TNode> = {|
-  added: Graph<TNode>,
-  removed: Graph<TNode>
-|};
-
-export interface Graph<TNode: Node> {
```

Comment:
These were removed as plugins no longer rely on these interfaces. `@parcel/types` is now only types belonging to the public api or types shared among parcel's own packages (which should overlap significantly)

Modules internal to core can still reference types from the concrete versions through e.g. `import type Graph from './Graph';`